plugins/de.cau.cs.kieler.klighd/src/de/cau/cs/kieler/klighd/filtering/parser/SemanticFilterRuleParser.java

```diff
@@ -0,0 +1,296 @@
+/*
+ * KIELER - Kiel Integrated Environment for Layout Eclipse RichClient
+ *
+ * http://rtsys.informatik.uni-kiel.de/kieler
+ * 
+ * Copyright 2022 by
+ * + Kiel University
+ *   + Department of Computer Science
+ *     + Real-Time and Embedded Systems Group
+ * 
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package de.cau.cs.kieler.klighd.filtering.parser;
+
+import static java.util.Map.entry;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.Stack;
+import java.util.StringTokenizer;
+
+import de.cau.cs.kieler.klighd.filtering.AndConnective;
+import de.cau.cs.kieler.klighd.filtering.FalseConnective;
+import de.cau.cs.kieler.klighd.filtering.GreaterEqualsConnective;
+import de.cau.cs.kieler.klighd.filtering.GreaterThanConnective;
+import de.cau.cs.kieler.klighd.filtering.LessEqualsConnective;
+import de.cau.cs.kieler.klighd.filtering.LessThanConnective;
+import de.cau.cs.kieler.klighd.filtering.LogicEqualConnective;
+import de.cau.cs.kieler.klighd.filtering.NegationConnective;
+import de.cau.cs.kieler.klighd.filtering.NumericAdditionConnective;
+import de.cau.cs.kieler.klighd.filtering.NumericConstantConnective;
+import de.cau.cs.kieler.klighd.filtering.NumericDivisionConnective;
+import de.cau.cs.kieler.klighd.filtering.NumericEqualConnective;
+import de.cau.cs.kieler.klighd.filtering.NumericMultiplicationConnective;
+import de.cau.cs.kieler.klighd.filtering.NumericNotEqualConnective;
+import de.cau.cs.kieler.klighd.filtering.NumericResult;
+import de.cau.cs.kieler.klighd.filtering.NumericSubtractionConnective;
+import de.cau.cs.kieler.klighd.filtering.OrConnective;
+import de.cau.cs.kieler.klighd.filtering.SemanticFilterRule;
+import de.cau.cs.kieler.klighd.filtering.SemanticFilterTag;
+import de.cau.cs.kieler.klighd.filtering.TrueConnective;
+
+/**
+ * @author mka
+ *
+ */
+public class SemanticFilterRuleParser {
+    
+    private abstract class Node {
+        public String token;
+        // current return type of AST
+        // -1: unknown
+        // 0: boolean
+        // 1: numeric
+        public int type;
+        public Node(String token, int type) {
+            this.token = token;
+            this.type = type;
+        }
+    }
+    
+    private class OperatorNode extends Node {
+        private List<Node> children;
+        
+        public List<Node> getChildren() { return children; }
+        
+        public OperatorNode(String token, int type) {
+            super(token, type);
+            this.children = new ArrayList<>();
+        }
+        public OperatorNode(String token, int type, List<Node> children) {
+            super(token, type);
+            this.children = children;
+        }
```

Comment:
This is always dangerous in Java to set the child list to the parameter and should be documented that no copy is made (which is fine here imo).